### PR TITLE
Add HTTP Proxy Support

### PR DESCRIPTION
**Pull request by zicklag**
Checklist:

Resolves #294.

* [x] `cargo +nightly fmt --all`
* [x] Modify `CHANGELOG.md` if applicable

### CODE REVIEW

**Comment by dpc** [comment]
Sweet. Needs some cleanup but I'd be happy to land such a PR.

File: Cargo.toml [hunk 1]
```diff
@@ -6,3 +6,7 @@ members = [
 	"crev-lib",
 	"crevsum"
 ]
+
+
+[patch.'crates-io']
```
Comment (dpc) [inline Cargo.toml:11]: Still needed?

Comment (zicklag) [reply Cargo.toml:11]: Still got to wait on https://github.com/theduke/crates_io_api/pull/19, or we could temporarily point it at the branch of my fork.

**Comment by zicklag** [comment]
I think this is working now, but I haven't successfully been able to add a trust to my crev-proofs while following the getting started guide. This is the first time I've gotten to use it; I'm not sure if I'm doing something wrong or if something else related to the proxy and/or the git clone is broken.

I will try again when I get a chance, but last I checked it was cloning the repo fine, but not checking out master so the attempt to commit the proof failed with some error about an unborn branch.

**Comment by dpc** [comment]
@zicklag Come join matrix/gitter channels in case of problems, and someone will help you. :)

**Comment by zicklag** [comment]
Thanks for the offer, but unfortunately network restrictions prevent me accessing chat. I should be able to figure it out, though, especially now that I've got the code all cracked open. :)

**Comment by dpc** [comment]
@zicklag In that case, feel free to just ping me here. I stare at my github notifications most of the time anyway.

**Comment by zicklag** [comment]
OK, I found the problem and updated some other missing updates for full proxy support.  That fixed the issues I had going through the getting started, so I'm glad about that! Other than the fact that we need to wait for the next release of `crates_io_api` I think this is ready.

Oh, other than the changelog. How should I notate a change to crev-lib? I bumped the crev-lib version, but that isn't the same version as in `cargo-crev`. Does `cargo-crev` get bumped as well and pointed at the new version of `crev-lib`?

**Comment by dpc** [comment]
@zicklag I am very lax about these stuff. Although the crates are split, it's mostly for keeping domain layers in order, and the whole thing behaves like a single project right now. Versions are kept the same everywhere, and there's only one CHANGELOG. `cargo-crev` is the main project, everything else is there to serve it.

There used to be more crates, but I've split out some, when it made sense and didn't introduce too much friction.

**Comment by zicklag** [comment]
OK, then I'll just make sure `crev-lib` and `cargo-crev` have the same version and put the record in the changelog under the new version.

**Comment by zicklag** [comment]
There you go, everything should be ready as soon as `crates_io_api` is published.

File: crev-lib/src/util/git.rs [hunk 2]
```diff
@@ -34,14 +35,40 @@ pub fn parse_git_url_https(http_url: &str) -> Option<GitUrlComponents> {
...
+    let fetch_options = default_fetch_options();
+    git2::build::RepoBuilder::new()
+        .fetch_options(fetch_options)
+        .clone(url, path.as_ref())
+}
+
+/// Get the default fetch options to use when fetching or cloneing
+///
+/// Currently this just ensures that git's automatic proxy settings are used.
+pub fn default_fetch_options<'a>() -> git2::FetchOptions<'a> {
```
Comment (dpc) [inline crev-lib/src/util/git.rs:62]: Wait, what? Where's this `<'a>` coming from? :D

File: crev-lib/src/util/mod.rs [hunk 3]
```diff
@@ -159,7 +164,7 @@ pub fn get_recursive_digest_for_paths(
     root_path: &Path,
     paths: fnv::FnvHashSet<PathBuf>,
 ) -> std::result::Result<Vec<u8>, crev_recursive_digest::DigestError> {
-    let h = crev_recursive_digest::RecursiveDigest::<crev_common::Blake2b256, _, _>::new()
+    let mut h = crev_recursive_digest::RecursiveDigest::<crev_common::Blake2b256, _, _>::new()
```
Comment (dpc) [inline crev-lib/src/util/mod.rs:167]: I think this `mut`s should be unnecessary now. You can try rebasing.

File: crev-lib/src/util/git.rs (hunk 2 above)
Comment (zicklag) [reply crev-lib/src/util/git.rs:62]: That is a good question, but for some strange reason the [`git2::FetchOptions`](https://docs.rs/git2/0.13.1/git2/struct.FetchOptions.html) acutally has a lifetime named `'dp` in it and I don't know where it comes from. If I don't put the `'a` in the function, Rust complains that there is nothing to borrow the `FetchOptions` from.

File: crev-lib/src/util/mod.rs (hunk 3 above)
Comment (zicklag) [reply crev-lib/src/util/mod.rs]: That worked.

**Comment by dpc** [comment]
How are we doing with this one?

**Comment by zicklag** [comment]
Ah, I had some strange access denied errors while I was attempting to finish off https://github.com/theduke/crates_io_api/pull/19. I'll look into it again when I can.

If anybody else is watching this and waiting on it tell me and I will prioritize it more ;).

**Comment by dpc** [comment]
No pressure, but I'll definitely get it landed. :)

**Comment by zicklag** [comment]
Ah, I just had to remove the `/` from the default user agent. Strange, but it was an easy fix.

Now we just have to wait for that PR to land in `crates_io_api`. I just pushed a rebased version of this PR that should be ready.

**Comment by dpc** [comment]
Oh noh. Conflicts. :(

**Comment by zicklag** [comment]
I'll wait until https://github.com/theduke/crates_io_api/pull/19 is merged then I'll resolve those. It shouldn't be a huge deal. I'll drop another ping over there too, to see if it can get merged.

**Comment by zicklag** [comment]
The `crates_io_api`  PR was merged so I'll try to take a look at these conflicts soon.

**Comment by tokcum** [comment]
> Thanks for the offer, but unfortunately network restrictions prevent me accessing chat. I should be able to figure it out, though, especially now that I've got the code all cracked open. :)

Same for me. I started with crev in my lab environment this week. I have Internet access via proxy but while working on the Getting Started with cargo-crev I stumbled across some issues related to Internet access.
The environment in the lab is quite restricted but Github is on the whitelist and I use it on a daily base.
I'm not sure if there are other, not whitelisted Internet ressources are involved or cargo-crev just does not fully support access proxy. It felt like the latter is the case but I didn't dive deep.

**Comment by zicklag** [comment]
Hey @tokcum I'll try to get to this hopefully in the next few days. Glad it will help somebody else, too. :smiley:

**Comment by tokcum** [comment]
> Hey @tokcum I'll try to get to this hopefully in the next few days. Glad it will help somebody else, too. smiley

I'm on vacations right now. In two weeks I'm back in the laboratory and are able to test it even if it still isn't merged.

**Comment by zicklag** [comment]
I resolved the conflicts, but there's a problem with this thread spawn because the `SyncClient` for `crates_io_api` isn't `Send` anymore. :/

https://github.com/crev-dev/cargo-crev/blob/7686a2f4eea18e9bd7b8eb021b214450b33f8de9/cargo-crev/src/deps/scan.rs#L171

@dpc you have any ideas for making that not need to be send over the thread or a place it would be fine to create a new client?

**Comment by dpc** [comment]
@zicklag 

https://github.com/theduke/crates_io_api/blob/41886283e1ac60a4356ace174edcccd0da542cac/src/sync_client.rs#L15

`RefCell` is `Send` https://doc.rust-lang.org/std/cell/struct.RefCell.html#impl-Send 

It just needs to be wrapped into `Arc<Mutex<_>>`` to be `Sync`, IIRC. 

If not, the author can be asked to turned that `RefCell` into `Mutex`.

**Comment by dpc** [comment]
Rebased and pushed as #437

**Pull request closed without merging**